File: pe/visualstudio/PE.JavaScript/Implementation/Problem0011.js
var javascriptPE;
(function (javascriptPE) {
    (function (implementation) {

        var problem0011 = (function() {

            function problem0011(listHelpers) {
                this.listHelper = listHelpers;
            }

            problem0011.prototype.maxProductOfHorizontals = function(inputArray) {
                var maximumProduct = 0;

                for (var i = 0; i < inputArray.length; i++) {
                    for (var j = 0; j < inputArray[i].length - 3; j++) {
                        var number1 = inputArray[i][j];
                        var number2 = inputArray[i][j + 1];
                        var number3 = inputArray[i][j + 2];
                        var number4 = inputArray[i][j + 3];

                        var product = number1 * number2 * number3 * number4;

                        if (product > maximumProduct) {
                            maximumProduct = product;
                        };
                    };
                };

                return maximumProduct;
            };

            problem0011.prototype.maxProductOfVerticles = function(inputArray) {
                var maximumProduct = 0;

                for (var i = 0; i < inputArray.length - 3; i++) {
                    for (var j = 0; j < inputArray[i].length; j++) {
                        var number1 = inputArray[i][j];
                        var number2 = inputArray[i + 1][j];
                        var number3 = inputArray[i + 2][j];
                        var number4 = inputArray[i + 3][j];

                        var product = number1 * number2 * number3 * number4;

                        if (product > maximumProduct) {
                            maximumProduct = product;
                        };
                    };
                };

                return maximumProduct;
            };

            problem0011.prototype.maxProductOfForwardDiagonals = function(inputArray) {
                var maximumProduct = 0;

                for (var i = 0; i < inputArray.length - 3; i++) {
                    for (var j = 3; j < inputArray[i].length; j++) {
                        var number1 = inputArray[i][j];
                        var number2 = inputArray[i + 1][j - 1];
                        var number3 = inputArray[i + 2][j - 2];
                        var number4 = inputArray[i + 3][j - 3];

                        var product = number1 * number2 * number3 * number4;

                        if (product > maximumProduct) {
                            maximumProduct = product;
                        };
                    };
                };

                return maximumProduct;
            };

            problem0011.prototype.maxProductOfBackwardDiagonals = function(inputArray) {
                var revArray = inputArray.reverse();
                return this.maxProductOfForwardDiagonals(revArray);
            };

            problem0011.prototype.maxProductOf4ConnectedNumbers = function(inputArray) {
                var horizontal = this.maxProductOfHorizontals(inputArray);
                var verticle = this.maxProductOfVerticles(inputArray);
                var forwards = this.maxProductOfForwardDiagonals(inputArray);
                var backwards = this.maxProductOfBackwardDiagonals(inputArray);

                var arrayOfResults = [horizontal, verticle, forwards, backwards];

                return this.listHelper.max(arrayOfResults);
            };


            return problem0011;
        })();
        implementation.problem0011 = problem0011;

    })(javascriptPE.implementation || (javascriptPE.implementation = {}));
})(javascriptPE || (javascriptPE = {}));

File: pe/visualstudio/PE.CSharp/Problem0011.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FSharp.MyLibraries;

namespace PE.CSharp
{
    public class Problem0011
    {
        public List<List<int>> ImportProblemSpace(string path)
        {
            var import = File.ReadAllLines(path);

            return import.Select(NumberParser.parseSpaceDelimitedString)
                    .Select(elem => elem.ToList())
                    .ToList();
        }

        public int MaxProductOfHorizontals(List<List<int>> inputMaxtrix)
        {
            var max = 0;

            foreach (var listOfInts in inputMaxtrix)
            {
                for (var i = 0; i < listOfInts.Count - 3; i++)
                {
                    var number1 = listOfInts[i];
                    var number2 = listOfInts[i + 1];
                    var number3 = listOfInts[i + 2];
                    var number4 = listOfInts[i + 3];

                    var product = number1 * number2 * number3 * number4;
                    if (product > max)
                    {
                        max = product;
                    }
                }
            }

            return max;
        }

        public int MaxProductOfVerticles(List<List<int>> inputMatrix)
        {
            var max = 0;

            for (var i = 0; i < inputMatrix.Count - 3; i++)
            {
                for (var j = 0; j < inputMatrix[i].Count; j++)
                {
                    var number1 = inputMatrix[i][j];
                    var number2 = inputMatrix[i + 1][j];
                    var number3 = inputMatrix[i + 2][j];
                    var number4 = inputMatrix[i + 3][j];

                    var product = number1 * number2 * number3 * number4;
                    if (product > max)
                    {
                        max = product;
                    }
                }
            }

            return max;
        }

        public int MaxProductOfForwardDiagonals(List<List<int>> inputMatrix)
        {
            var max = 0;

            for (var i = 0; i < inputMatrix.Count - 3; i++)
            {
                for (var j = 3; j < inputMatrix[i].Count; j++)
                {
                    var number1 = inputMatrix[i][j];
                    var number2 = inputMatrix[i + 1][j - 1];
                    var number3 = inputMatrix[i + 2][j - 2];
                    var number4 = inputMatrix[i + 3][j - 3];

                    var product = number1 * number2 * number3 * number4;
                    if (product > max)
                    {
                        max = product;
                    }
                }
            }

            return max;
        }

        public int MaxProductOfBackwardDiagonals(List<List<int>> inputMatrix)
        {
            inputMatrix.Reverse();

            return MaxProductOfForwardDiagonals(inputMatrix);
        }

        public int MaxProductOf4ConnectedNumbers(string inputPath)
        {
            var inputMatrix = ImportProblemSpace(inputPath);

            var listOfProducts = new List<int>
            {
                MaxProductOfVerticles(inputMatrix),
                MaxProductOfHorizontals(inputMatrix),
                MaxProductOfForwardDiagonals(inputMatrix),
                MaxProductOfBackwardDiagonals(inputMatrix)
            };

            return listOfProducts.Max();
        }
    }
}
